File: src/game/scoring.ts
import type { JudgeResult, Outfit, RGB, Theme } from './types';

const judges = ['Regina', 'Gretchen', 'Karen'] as const;

export function scoreOutfit(theme: Theme, outfit: Outfit): { finalScore: number; judges: JudgeResult[] } {
  const items = Object.values(outfit);

  if (items.length === 0) {
    return {
      finalScore: 10,
      judges: judges.map((name) => ({ name, score: 3, comment: 'Needs more styling.' })),
    };
  }

  const allTags = items.flatMap((item) => item.tags);
  const matchedTags = allTags.filter((tag) => theme.tags.includes(tag)).length;
  const colorHits = items.filter((item) => theme.colors.some((color) => colorDistance(item.color, color) < 150)).length;
  const categoryBonus = items.length * 4;
  const varietyBonus = new Set(items.map((item) => item.shape)).size * 2;
  const base = clamp(12 + matchedTags * 7 + colorHits * 5 + categoryBonus + varietyBonus, 1, 100);

  const judgeResults = judges.map((name, index) => {
    const wobble = deterministicWobble(theme.name, allTags.join('-'), index);
    const score = clamp(Math.round(base + wobble), 1, 100);
    return {
      name,
      score,
      comment: getComment(score),
    };
  });

  const finalScore = Math.round(judgeResults.reduce((total, judge) => total + judge.score, 0) / judgeResults.length);
  return { finalScore, judges: judgeResults };
}

function colorDistance(a: RGB, b: RGB): number {
  return Math.abs(a[0] - b[0]) + Math.abs(a[1] - b[1]) + Math.abs(a[2] - b[2]);
}

function deterministicWobble(themeName: string, outfitKey: string, index: number): number {
  const text = `${themeName}-${outfitKey}-${index}`;
  const total = [...text].reduce((sum, char) => sum + char.charCodeAt(0), 0);
  return (total % 17) - 8;
}

function getComment(score: number): string {
  if (score >= 85) return 'Iconic theme match!';
  if (score >= 65) return 'Cute, clear, and stylish.';
  if (score >= 45) return 'Good start, add stronger theme details.';
  return 'Try more matching pieces.';
}

function clamp(value: number, min: number, max: number): number {
  return Math.min(max, Math.max(min, value));
}

File: style_rush/scoring.py
import random
from dataclasses import dataclass

from themes import Theme
from wardrobe import ClothingItem


@dataclass
class JudgeResult:
    name: str
    score: int
    comment: str


JUDGES = ("Regina", "Gretchen", "Karen")


def color_distance(a: tuple[int, int, int], b: tuple[int, int, int]) -> int:
    return abs(a[0] - b[0]) + abs(a[1] - b[1]) + abs(a[2] - b[2])


def score_outfit(theme: Theme, outfit: dict[str, ClothingItem]) -> tuple[int, list[JudgeResult]]:
    if not outfit:
        return 10, [JudgeResult(name, 3, "Needs more styling.") for name in JUDGES]

    all_tags = [tag for item in outfit.values() for tag in item.tags]
    matched_tags = sum(1 for tag in all_tags if tag in theme.tags)
    color_hits = 0
    for item in outfit.values():
        if min(color_distance(item.color, target) for target in theme.colors) < 150:
            color_hits += 1

    category_bonus = len(outfit) * 4
    variety_bonus = len(set(item.shape for item in outfit.values())) * 2
    theme_score = matched_tags * 7 + color_hits * 5
    base = min(100, 12 + category_bonus + variety_bonus + theme_score)

    results: list[JudgeResult] = []
    for judge in JUDGES:
        score = max(1, min(100, base + random.randint(-8, 8)))
        if score >= 85:
            comment = "Iconic theme match!"
        elif score >= 65:
            comment = "Cute, clear, and stylish."
        elif score >= 45:
            comment = "Good start, needs stronger theme details."
        else:
            comment = "Try adding more matching pieces."
        results.append(JudgeResult(judge, score, comment))

    average = round(sum(result.score for result in results) / len(results))
    return average, results
